File: Task/Yellowstone-sequence/Go/yellowstone-sequence.go
package main

import (
    "fmt"
    "log"
    "os/exec"
)

func gcd(x, y int) int {
    for y != 0 {
        x, y = y, x%y
    }
    return x
}

func yellowstone(n int) []int {
    m := make(map[int]bool)
    a := make([]int, n+1)
    for i := 1; i < 4; i++ {
        a[i] = i
        m[i] = true
    }
    min := 4
    for c := 4; c <= n; c++ {
        for i := min; ; i++ {
            if !m[i] && gcd(a[c-1], i) == 1 && gcd(a[c-2], i) > 1 {
                a[c] = i
                m[i] = true
                if i == min {
                    min++
                }
                break
            }
        }
    }
    return a[1:]
}

func check(err error) {
    if err != nil {
        log.Fatal(err)
    }
}

func main() {
    x := make([]int, 100)
    for i := 0; i < 100; i++ {
        x[i] = i + 1
    }
    y := yellowstone(100)
    fmt.Println("The first 30 Yellowstone numbers are:")
    fmt.Println(y[:30])
    g := exec.Command("gnuplot", "-persist")
    w, err := g.StdinPipe()
    check(err)
    check(g.Start())
    fmt.Fprintln(w, "unset key; plot '-'")
    for i, xi := range x {
        fmt.Fprintf(w, "%d %d\n", xi, y[i])
    }
    fmt.Fprintln(w, "e")
    w.Close()
    g.Wait()
}

File: Task/Yellowstone-sequence/Zen-C/yellowstone-sequence.c
import "std/set.zc"

fn gcd(x: int, y: int) -> int {
    while y {
        let t = y;
        y = x % y;
        x = t;
    }
    return x >= 0 ? x :-x;
}

fn yellowstone(n: int, a: int*) {
    let s = Set<int>::new();
    for i in 1..4 {
        a[i] = i;
        s.add(i);
    }
    let min = 4;
    for c in 4..=n {
        for let i = min; ; ++i {
            if !s.contains(i) && gcd(a[c - 1], i) == 1 && gcd(a[c - 2], i) > 1 {
                a[c] = i;
                s.add(i);
                if i == min { min++; }
                break;
            }
        }
    }
}

fn main() {
    let x: int[30];
    for i in 0..30 { x[i] = i + 1; }
    let a: [int; 31];
    yellowstone(30, (int*)a);
    println "The first 30 Yellowstone numbers are:";
    for i in 1..=30 {
        print "{a[i]:2d}  ";
        if !(i % 10) { println ""; }
    }
}
